File: verenigingen/verenigingen/page/chapter_dues_allocation/chapter_dues_allocation.py
"""
Chapter Dues Allocation Page Backend

Server-side methods for generating journal entries to allocate dues income
between chapter and national accounts.
"""

from typing import Dict, List, Optional

import frappe
from frappe import _
from frappe.utils import flt, getdate, nowdate

from verenigingen.utils.security.api_security_framework import OperationType, critical_api, high_security_api
from verenigingen.verenigingen.domain.chapter_dues import DuesAllocationService
from verenigingen.verenigingen.domain.chapter_dues_validation import DuesAllocationValidator


@frappe.whitelist()
@high_security_api(operation_type=OperationType.FINANCIAL)
def get_allocation_preview(
    from_date: str, to_date: str, chapter: Optional[str] = None, company: Optional[str] = None
) -> Dict:
    """
    Get preview of allocation amounts before generating journal entries.

    Args:
        from_date: Start date for invoice filtering
        to_date: End date for invoice filtering
        chapter: Optional chapter filter
        company: Optional company filter

    Returns:
        Dictionary with allocation preview data
    """
    # Validate inputs using centralized validator
    validator = DuesAllocationValidator()

    try:
        from_date = getdate(from_date)
        to_date = getdate(to_date)
    except Exception:
        frappe.throw(_("Invalid date format"))

    validator.validate_date_range(from_date, to_date)
    validator.validate_chapter_exists(chapter)
    validator.validate_company_exists(company)

    # Get settings for account configuration
    settings = frappe.get_single("Verenigingen Settings")

    # Build query conditions
    conditions = ["si.docstatus = 1", "si.custom_member_chapter IS NOT NULL"]
    filters = {"from_date": from_date, "to_date": to_date}

    if chapter:
        conditions.append("si.custom_member_chapter = %(chapter)s")
        filters["chapter"] = chapter
    if company:
        conditions.append("si.company = %(company)s")
        filters["company"] = company

    where_clause = " AND ".join(conditions)

    # Query chapter totals
    query = f"""
        SELECT
            si.custom_member_chapter as chapter,
            si.company,
            COUNT(si.name) as invoice_count,
            SUM(si.grand_total) as total_amount
        FROM `tabSales Invoice` si
        WHERE {where_clause}
            AND si.posting_date >= %(from_date)s
            AND si.posting_date <= %(to_date)s
        GROUP BY si.custom_member_chapter, si.company
        ORDER BY si.custom_member_chapter
    """

    chapter_data = frappe.db.sql(query, filters, as_dict=True)

    # Use domain service for calculations (avoids N+1 queries)
    allocation_service = DuesAllocationService()

    # Prepare batch input
    chapter_amounts = {row["chapter"]: flt(row["total_amount"]) for row in chapter_data}

    # Batch calculate all allocations (single query for all chapter configs)
    allocations_map = allocation_service.batch_calculate(chapter_amounts)

    # Build response
    allocations = []
    total_chapter_amount = 0
    total_national_amount = 0

    for row in chapter_data:
        chapter_name = row["chapter"]
        allocation = allocations_map[chapter_name]

        allocations.append(
            {
                "chapter": chapter_name,
                "company": row["company"],
                "invoice_count": row["invoice_count"],
                **allocation.to_dict(),
            }
        )

        total_chapter_amount += float(allocation.chapter_amount)
        total_national_amount += float(allocation.national_amount)

    return {
        "allocations": allocations,
        "summary": {
            "total_chapter_amount": round(total_chapter_amount, 2),
            "total_national_amount": round(total_national_amount, 2),
            "grand_total": round(total_chapter_amount + total_national_amount, 2),
        },
        "accounts": {
            "chapter_account": settings.chapter_dues_income_account,
            "national_account": settings.national_dues_income_account,
            "source_account": settings.dues_income_account,
        },
    }


@frappe.whitelist()
@critical_api(operation_type=OperationType.FINANCIAL)
def generate_allocation_journal_entries(
    from_date: str,
    to_date: str,
    posting_date: Optional[str] = None,
    chapter: Optional[str] = None,
    company: Optional[str] = None,
) -> Dict:
    """
    Generate Journal Entries to allocate dues income between chapter and national accounts.

    Args:
        from_date: Start date for invoice filtering
        to_date: End date for invoice filtering
        posting_date: Posting date for journal entries (defaults to today)
        chapter: Optional chapter filter
        company: Optional company filter

    Returns:
        List of created Journal Entry names
    """
    # Validate inputs using centralized validator
    validator = DuesAllocationValidator()

    try:
        from_date = getdate(from_date)
        to_date = getdate(to_date)
        if posting_date:
            posting_date = getdate(posting_date)
        else:
            posting_date = nowdate()
    except Exception:
        frappe.throw(_("Invalid date format"))

    validator.validate_date_range(from_date, to_date)
    validator.validate_chapter_exists(chapter)
    validator.validate_company_exists(company)

    # Check for idempotency - prevent duplicate allocations
    # Build allocation period key for tracking
    allocation_period_key = f"{from_date}|{to_date}|{chapter or 'ALL'}|{company or 'ALL'}"

    existing_entries = frappe.get_all(
        "Journal Entry",
        filters={
            "custom_dues_allocation_period": allocation_period_key,
            "docstatus": ["!=", 2],  # Not cancelled
        },
        fields=["name", "docstatus"],
    )

    if existing_entries:
        entry_list = ", ".join([e.name for e in existing_entries])
        frappe.throw(
            _(
                "Allocation journal entries already exist for period {0} to {1}: {2}. "
                "Please cancel existing entries before creating new ones."
            ).format(from_date, to_date, entry_list),
            title=_("Duplicate Allocation Detected"),
        )

    # Get allocation preview
    preview = get_allocation_preview(from_date, to_date, chapter, company)

    if not preview["allocations"]:
        frappe.throw(_("No dues invoices found for the selected period and filters"))

    # Validate account configuration using centralized validator
    validator.validate_account_configuration()

    # Get settings for account access
    settings = frappe.get_single("Verenigingen Settings")

    created_entries = []

    # Wrap journal entry creation in try-except for proper rollback on partial failure
    try:
        # Create one Journal Entry per company
        for company_name in set([a["company"] for a in preview["allocations"]]):
            company_allocations = [a for a in preview["allocations"] if a["company"] == company_name]

            # Calculate company totals
            company_chapter_total = sum([a["chapter_amount"] for a in company_allocations])
            company_national_total = sum([a["national_amount"] for a in company_allocations])
            company_grand_total = company_chapter_total + company_national_total

            # Create Journal Entry
            #
            # Accounting Logic:
            # -----------------
            # This journal entry redistributes dues income from a single source account
            # into separate chapter and national accounts based on configured split percentages.
            #
            # Double-Entry Structure:
            # 1. Debit (Dr) Dues Income Account:     Total amount (reverses original income)
            # 2. Credit (Cr) Chapter Income Account: Chapter portion (records split income)
            # 3. Credit (Cr) National Income Account: National portion (records split income)
            #
            # Example for €100 dues with 60/40 split:
            # Dr Dues Income Account        €100
            #    Cr Chapter Income Account           €60
            #    Cr National Income Account          €40
            #
            # This maintains the accounting equation: Assets = Liabilities + Equity
            # The total income remains unchanged, only its classification is adjusted.
            # Build allocation period key for this company
            company_allocation_key = f"{from_date}|{to_date}|{chapter or 'ALL'}|{company_name}"

            je = frappe.get_doc(
                {
                    "doctype": "Journal Entry",
                    "voucher_type": "Journal Entry",
                    "company": company_name,
                    "posting_date": posting_date,
                    "user_remark": f"Allocation of membership dues income for period {from_date} to {to_date}",
                    "custom_dues_allocation_period": company_allocation_key,
                    "accounts": [
                        {
                            "account": settings.dues_income_account,
                            "debit_in_account_currency": company_grand_total,
                            "credit_in_account_currency": 0,
                            "user_remark": f"Reverse total dues income",
                        },
                        {
                            "account": settings.chapter_dues_income_account,
                            "debit_in_account_currency": 0,
                            "credit_in_account_currency": company_chapter_total,
                            "user_remark": f"Allocate chapter portion ({len(company_allocations)} chapters)",
                        },
                        {
                            "account": settings.national_dues_income_account,
                            "debit_in_account_currency": 0,
                            "credit_in_account_currency": company_national_total,
                            "user_remark": f"Allocate national portion",
                        },
                    ],
                }
            )

            # Validate accounting entries balance
            total_debits = sum([acc.get("debit_in_account_currency", 0) for acc in je.accounts])
            total_credits = sum([acc.get("credit_in_account_currency", 0) for acc in je.accounts])

            if round(total_debits, 2) != round(total_credits, 2):
                frappe.throw(
                    _("Journal Entry accounting error for {0}: Debits ({1}) != Credits ({2})").format(
                        company_name, total_debits, total_credits
                    )
                )

            je.insert()
            created_entries.append(je.name)

            frappe.msgprint(
                _("Created Journal Entry {0} for {1}").format(frappe.bold(je.name), frappe.bold(company_name))
            )

    except Exception as e:
        # Rollback all changes if any journal entry creation fails
        frappe.db.rollback()
        frappe.log_error(
            title=_("Journal Entry Creation Failed"),
            message=f"Failed to create journal entries for period {from_date} to {to_date}. "
            f"Created {len(created_entries)} entries before failure. Error: {str(e)}",
        )
        frappe.throw(
            _(
                "Journal entry creation failed after creating {0} of {1} entries. "
                "All changes have been rolled back. Error: {2}"
            ).format(len(created_entries), len(set([a["company"] for a in preview["allocations"]])), str(e)),
            title=_("Transaction Rolled Back"),
        )

    # Frappe will auto-commit when API request completes successfully
    # Manual commit removed to allow proper rollback on errors

    return {
        "success": True,
        "journal_entries": created_entries,
        "message": _("Successfully created {0} Journal Entry(s)").format(len(created_entries)),
    }

File: verenigingen/verenigingen/page/chapter_dues_allocation/chapter_dues_allocation.js
frappe.pages['chapter-dues-allocation'].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Chapter Dues Allocation',
		single_column: true
	});

	page.allocation_tool = new ChapterDuesAllocationTool(page);
};

class ChapterDuesAllocationTool {
	constructor(page) {
		this.page = page;
		this.make();
	}

	make() {
		this.make_form();
		this.make_result_area();
		this.setup_actions();
	}

	make_form() {
		this.form_container = $('<div class="frappe-control">').appendTo(this.page.main);

		this.from_date = frappe.ui.form.make_control({
			parent: this.form_container,
			df: {
				fieldtype: 'Date',
				label: __('From Date'),
				fieldname: 'from_date',
				default: frappe.datetime.month_start(),
				reqd: 1
			},
			render_input: true
		});

		this.to_date = frappe.ui.form.make_control({
			parent: this.form_container,
			df: {
				fieldtype: 'Date',
				label: __('To Date'),
				fieldname: 'to_date',
				default: frappe.datetime.month_end(),
				reqd: 1
			},
			render_input: true
		});

		this.chapter = frappe.ui.form.make_control({
			parent: this.form_container,
			df: {
				fieldtype: 'Link',
				label: __('Chapter (Optional)'),
				fieldname: 'chapter',
				options: 'Chapter'
			},
			render_input: true
		});

		this.company = frappe.ui.form.make_control({
			parent: this.form_container,
			df: {
				fieldtype: 'Link',
				label: __('Company'),
				fieldname: 'company',
				options: 'Company',
				default: frappe.defaults.get_user_default('Company')
			},
			render_input: true
		});

		this.posting_date = frappe.ui.form.make_control({
			parent: this.form_container,
			df: {
				fieldtype: 'Date',
				label: __('Journal Entry Posting Date'),
				fieldname: 'posting_date',
				default: frappe.datetime.get_today(),
				reqd: 1
			},
			render_input: true
		});
	}

	make_result_area() {
		this.result_area = $('<div class="result-area" style="margin-top: 20px;">').appendTo(this.page.main);
	}

	setup_actions() {
		const me = this;

		this.page.set_primary_action(__('Preview Allocation'), () => {
			me.preview_allocation();
		});

		this.page.add_inner_button(__('Generate Journal Entries'), () => {
			me.generate_journal_entries();
		}, __('Actions'));

		this.page.add_inner_button(__('View Chapter Dues Split Report'), () => {
			frappe.set_route('query-report', 'Chapter Dues Split');
		}, __('Reports'));
	}

	get_filters() {
		return {
			from_date: this.from_date.get_value(),
			to_date: this.to_date.get_value(),
			chapter: this.chapter.get_value(),
			company: this.company.get_value()
		};
	}

	preview_allocation() {
		const me = this;
		const filters = this.get_filters();

		if (!filters.from_date || !filters.to_date) {
			frappe.msgprint(__('Please select From Date and To Date'));
			return;
		}

		// Validate date range
		if (filters.from_date > filters.to_date) {
			frappe.msgprint({
				title: __('Invalid Date Range'),
				message: __('From Date cannot be after To Date'),
				indicator: 'red'
			});
			return;
		}

		frappe.call({
			method: 'verenigingen.verenigingen.page.chapter_dues_allocation.chapter_dues_allocation.get_allocation_preview',
			args: filters,
			freeze: true,
			freeze_message: __('Loading allocation preview...'),
			callback(r) {
				if (r.message) {
					try {
						me.render_preview(r.message);
					} catch (error) {
						console.error('Error rendering preview:', error);
						frappe.msgprint({
							title: __('Rendering Error'),
							message: __('Failed to display allocation preview. Check browser console for details.'),
							indicator: 'red'
						});
					}
				} else {
					frappe.msgprint({
						title: __('No Data'),
						message: __('No allocation data returned from server'),
						indicator: 'orange'
					});
				}
			},
			error(r) {
				console.error('API error:', r);

				// Try to extract error message from response
				let error_message = __('Failed to load allocation preview');
				if (r && r.message) {
					error_message = r.message;
				} else if (r && r._server_messages) {
					try {
						const messages = JSON.parse(r._server_messages);
						if (messages && messages.length > 0) {
							const parsed = JSON.parse(messages[0]);
							error_message = parsed.message || error_message;
						}
					} catch (e) {
						// Ignore JSON parse errors
					}
				}

				frappe.msgprint({
					title: __('Error'),
					message: error_message,
					indicator: 'red'
				});
			}
		});
	}

	render_preview(data) {
		// Validate data structure
		if (!data || typeof data !== 'object') {
			throw new Error('Invalid data structure: expected object');
		}

		if (!Array.isArray(data.allocations)) {
			throw new Error('Invalid data structure: allocations must be an array');
		}

		let html = '<div class="allocation-preview">';
		html += `<h4>${__('Allocation Preview')}</h4>`;

		if (data.allocations.length === 0) {
			html += `<p class="text-muted">${__('No dues invoices found for the selected period')}</p>`;
			this.result_area.html(`${html}</div>`);
			return;
		}

		// Validate accounts object exists
		if (!data.accounts || typeof data.accounts !== 'object') {
			console.warn('Missing accounts data in preview response');
			data.accounts = {};
		}

		// Account configuration warning
		if (!data.accounts.chapter_account || !data.accounts.national_account) {
			html += '<div class="alert alert-warning">';
			html += `<strong>${__('Warning')}:</strong> `;
			html += __('Chapter and National Dues Income Accounts are not configured in Verenigingen Settings. ');
			html += __('Please configure these before generating journal entries.');
			html += '</div>';
		}

		// Summary table
		html += '<table class="table table-bordered">';
		html += '<thead><tr>';
		html += `<th>${__('Chapter')}</th>`;
		html += `<th class="text-right">${__('Invoices')}</th>`;
		html += `<th class="text-right">${__('Total Dues')}</th>`;
		html += `<th class="text-right">${__('Chapter %')}</th>`;
		html += `<th class="text-right">${__('Chapter Amount')}</th>`;
		html += `<th class="text-right">${__('National Amount')}</th>`;
		html += '</tr></thead><tbody>';

		data.allocations.forEach((row) => {
			html += '<tr>';
			html += `<td>${row.chapter}</td>`;
			html += `<td class="text-right">${row.invoice_count}</td>`;
			html += `<td class="text-right">${format_currency(row.total_amount)}</td>`;
			html += `<td class="text-right">${row.chapter_percentage.toFixed(1)}%</td>`;
			html += `<td class="text-right">${format_currency(row.chapter_amount)}</td>`;
			html += `<td class="text-right">${format_currency(row.national_amount)}</td>`;
			html += '</tr>';
		});

		// Totals row
		html += '<tr class="font-weight-bold">';
		html += `<td colspan="4">${__('Grand Total')}</td>`;
		html += `<td class="text-right">${format_currency(data.summary.total_chapter_amount)}</td>`;
		html += `<td class="text-right">${format_currency(data.summary.total_national_amount)}</td>`;
		html += '</tr>';

		html += '</tbody></table>';

		// Account info
		if (data.accounts.chapter_account && data.accounts.national_account) {
			html += '<div class="mt-3">';
			html += `<h5>${__('Journal Entry Accounts')}</h5>`;
			html += '<ul>';
			html += `<li><strong>${__('Source')}:</strong> ${data.accounts.source_account || __('Not configured')}</li>`;
			html += `<li><strong>${__('Chapter Account')}:</strong> ${data.accounts.chapter_account}</li>`;
			html += `<li><strong>${__('National Account')}:</strong> ${data.accounts.national_account}</li>`;
			html += '</ul>';
			html += '</div>';
		}

		html += '</div>';

		this.result_area.html(html);
	}

	generate_journal_entries() {
		const me = this;
		const filters = this.get_filters();
		filters.posting_date = this.posting_date.get_value();

		if (!filters.from_date || !filters.to_date || !filters.posting_date) {
			frappe.msgprint(__('Please fill all required fields'));
			return;
		}

		// Validate date range
		if (filters.from_date > filters.to_date) {
			frappe.msgprint({
				title: __('Invalid Date Range'),
				message: __('From Date cannot be after To Date'),
				indicator: 'red'
			});
			return;
		}

		// Validate posting date is not in future (warn only)
		const today = frappe.datetime.get_today();
		if (filters.posting_date > today) {
			frappe.msgprint({
				title: __('Future Posting Date'),
				message: __('Warning: Posting date is in the future. This is allowed but unusual.'),
				indicator: 'orange'
			});
		}

		frappe.confirm(
			__('This will create Journal Entry(s) to allocate membership dues income. Continue?'),
			() => {
				frappe.call({
					method: 'verenigingen.verenigingen.page.chapter_dues_allocation.chapter_dues_allocation.generate_allocation_journal_entries',
					args: filters,
					freeze: true,
					freeze_message: __('Creating journal entries...'),
					callback(r) {
						if (r.message && r.message.success) {
							frappe.msgprint({
								title: __('Success'),
								message: r.message.message,
								indicator: 'green'
							});

							// Show links to created journal entries
							if (r.message.journal_entries && r.message.journal_entries.length > 0) {
								let msg = `${__('Created Journal Entries')}:<br>`;
								r.message.journal_entries.forEach((name) => {
									msg += `<a href="/app/journal-entry/${name}">${name}</a><br>`;
								});
								frappe.msgprint(msg);
							}

							// Refresh preview
							try {
								me.preview_allocation();
							} catch (error) {
								console.error('Error refreshing preview:', error);
								// Don't show error to user - journal entries were created successfully
							}
						} else if (r.message && !r.message.success) {
							// Server returned response but operation failed
							frappe.msgprint({
								title: __('Operation Failed'),
								message: r.message.message || __('Failed to create journal entries'),
								indicator: 'red'
							});
						}
					},
					error(r) {
						console.error('API error:', r);

						// Try to extract error message from response
						let error_message = __('Failed to create journal entries');
						if (r && r.message) {
							error_message = r.message;
						} else if (r && r._server_messages) {
							try {
								const messages = JSON.parse(r._server_messages);
								if (messages && messages.length > 0) {
									const parsed = JSON.parse(messages[0]);
									error_message = parsed.message || error_message;
								}
							} catch (e) {
								// Ignore JSON parse errors
							}
						}

						frappe.msgprint({
							title: __('Error'),
							message: error_message,
							indicator: 'red'
						});
					}
				});
			}
		);
	}
}
